src/storage/src/hummock/compactor.rs

```diff
@@ -147,10 +147,59 @@ impl Compactor {
         }
     }
 
+    /// Flush shared buffer to level0. Resulted SSTs are grouped by compaction group.
+    pub async fn compact_shared_buffer_by_compaction_group(
+        context: Arc<CompactorContext>,
+        payload: UploadTaskPayload,
+    ) -> HummockResult<Vec<(CompactionGroupId, Sstable, u64, Vec<VNodeBitmap>)>> {
+        let mut grouped_payload: HashMap<CompactionGroupId, UploadTaskPayload> = HashMap::new();
+        for uncommitted_list in payload {
+            let mut next_inner = HashSet::new();
+            for uncommitted in uncommitted_list {
+                let batch = match &uncommitted {
+                    UncommittedData::Sst(_) => {
+                        panic!("expect UncommittedData::Batch")
+                    }
+                    UncommittedData::Batch(batch) => batch,
+                };
+                let group = grouped_payload
+                    .entry(batch.compaction_group_id())
+                    .or_insert_with(std::vec::Vec::new);
+                if !next_inner.contains(&batch.compaction_group_id()) {
+                    group.push(vec![]);
+                    next_inner.insert(batch.compaction_group_id());
+                }
+                group.last_mut().unwrap().push(uncommitted);
```

Comment:
If we "do the compaction group separation in SharedBuffer::new_upload_task and submit one upload task per compaction group", the key in `uploading_tasks: HashMap<OrderIndex, KeyIndexedUncommittedData>,` should be changed to (OrderIndex, CompactionGroupId) as well, is it? I think it will complicate the code to some extent? @wenym1

Reply:
Agree with @zwang28 . We can merge this PR and I will refine the logic of new upload task providing compact group id.